test/resources/knativeserving.go

```diff
@@ -59,22 +119,28 @@ func WaitForKnativeServingState(client *test.KnativeServingAlphaClients, name st
 }
 
 // IsKnativeServingReady will check the status conditions of the KnativeServing and return true if the KnativeServing is ready.
-func IsKnativeServingReady(s *v1alpha1.KnativeServing) (bool, error) {
+func IsKnativeServingReady(s *v1alpha1.KnativeServing, err error) (bool, error) {
 	return s.Status.IsReady(), nil
 }
 
-// WaitForDeploymentAvailable polls the status of the deployment called name
+// IsKnativeServingDeleted will check the status conditions of the KnativeServing and return true if the KnativeServing is deleted.
+func IsKnativeServingDeleted(s *v1alpha1.KnativeServing, err error) (bool, error) {
+	return apierrs.IsNotFound(err), nil
+}
+
+// WaitForDeployment polls the status of the deployment called name
 // from client every `interval` until `inState` returns `true` indicating it
 // is done, returns an error or timeout.
-func WaitForDeploymentAvailable(clients *test.Clients, name, namespace string, inState func(s *v1.Deployment) (bool, error)) (*v1.Deployment, error) {
+func WaitForDeployment(clients *test.Clients, name, namespace string, inState func(s *v1.Deployment, err error) (bool, error),
```

Comment:
Why not use the upstream `pkg/test/WaitForDeploymentState`?

Reply:
good catch. will use it.

Reply:
I don't understand. You seem to be adding the error arg, so that would enable you to use the existing one, right?

Reply:
After adding error param, cannot use the existing one in pkg/test/WaitForDeploymentState.

Reply:
Gotcha. So that's even easier. Your `inState` can just always return false, because it won't even get called at all if everything works as expected. You'll just need to do your `IsNotFound` check in the test calling `WaitForDeploymentState` because it will be returning the error you expect.

Reply:
I use this func WaitForDeployment in two situations: one to wait until the deployment is available, the other is to wait until it is removed. For the first one I need to check the deployment itself, but for the second one I need to check if error is NotFound, indicating deployment is gone.  For the instate func to verify the status, I just take both of the deployment and error as params, so WaitForDeployment can resued for these two cases. It has to be different from pkg/test/WaitForDeploymentState.

Reply:
I understand. All I'm suggesting is that you can simplify our code by using the existing `pkg/test` helper, removing the `error` arg from your instate functions, having `isDeploymentDeleted` always return false and changing your fatal failure condition in `TestKnativeServingDeletion` to something like this:
```
	for _, deployment := range dpListSave.Items {
		if _, err := test.WaitForDeploymentState(..., resources.IsDeploymentDeleted, ...); !apierrs.IsNotFound(err) {
			t.Fatalf(...)
		}
		t.Logf(...)
```
Make sense?

Reply:
if resources.IsDeploymentDeleted always return false, WaitForDeploymentState will never get the desired condition, before timeout. 

when IsDeploymentDeleted return true, WaitForDeploymentState meet the desired condition to quite.

Reply:
I think WaitForDeploymentState can be used to check if deployment is removed,
but for waiting deployment till available, WaitForDeploymentState will return error, if deployment is not found.
I still do not know how to use it for this case.
During deployment recreation, it is possible that deployment is not available, but will come back later.